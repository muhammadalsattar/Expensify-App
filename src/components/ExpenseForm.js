import React from "react";
import moment from "moment";
import {SingleDatePicker} from 'react-dates';
import { connect } from "react-redux";

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            description: props.expense? props.expense.description:'',
            note: props.expense? props.expense.note:'',
            amount: props.expense? (props.expense.amount / 100).toString():'',
            createdAt: props.expense? moment(props.expense.createdAt): moment(),
            focused: false,
            error: ''
        }
    }
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState((state) => ({description}))
    }
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState((state) => ({note}))
    }
    onAmountChange = (e) => {
        const amount = e.target.value
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState((state) => ({amount}))
        }
    }
    onDateChange = (date) => {
        if(date){
            this.setState(() => ({createdAt:date}))
        }
    }
    onFocusChange = ({focused}) => {
        this.setState(() => ({focused}))
    }
    onSubmit = (e) => {
        e.preventDefault()

        if (!this.state.description || !this.state.amount)
        {
            this.setState(() => ({error: 'Please provide description and amount!'}))
            if (e.target){
                e.target.parentNode.style.display='block';
            }
        }
        else{
            this.setState((state) => ({error: ''}))
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    }
    closeAlert = ()=>{
        const alertNode = document.querySelector('.alert')
        alertNode.style.display='none';
    }
    render(){
        return(
            <div className="expense-form">
                {this.state.error && <div className="alert mx-auto alert-danger alert-dismissible fade show" role="alert">
                {this.state.error}
                <button type="button" onClick={this.closeAlert} className="btn-close" data-bs-dismiss="alert" aria-label="Close"></button>
              </div>}
                <form className="mx-auto my-4" onSubmit={this.onSubmit}>
                        <input className="form-control"  type="text" placeholder="description" autoFocus value={this.state.description} onChange={this.onDescriptionChange}></input>
                        <input className="form-control" type="text" placeholder="Amount" value={this.state.amount} onChange={this.onAmountChange}></input>
                        <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.focused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => (false)}
                        />
                    <textarea className="form-control" placeholder="Add note for this expense" value={this.state.note} onChange={this.onNoteChange}></textarea>
                    <button className="btn btn-dark submit-button">Save Expense</button>
                </form>
            </div>
        )
    }
}

const ConnectedExpenseForm = connect()(ExpenseForm)