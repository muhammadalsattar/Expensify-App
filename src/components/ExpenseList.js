import React from "react";
import { connect } from "react-redux";
import ConnectedExpenseItem from "./ExpenseListItem";
import getVisibleExpenses from "../selectors/expenses";

export function ExpenseList(props) {
    return (
        <div>
        {
            props.expenses.length < 1? (
                <p>No Expenses yet</p>
            ):(
                props.expenses.map((expense => {
                    return <ConnectedExpenseItem {...expense} key={expense.id}></ConnectedExpenseItem>;
                }))
            )
        }
        </div>
    );
}

const MapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    }
}

const ConnectedExpenseList = connect(MapStateToProps)(ExpenseList)

export default ConnectedExpenseList;