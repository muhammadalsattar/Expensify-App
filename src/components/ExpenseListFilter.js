import React from "react";
import { connect } from "react-redux";
import { setFilterText } from "../actions/filters";
import { sortByAmount, sortByDate } from "../actions/filters";
import {DateRangePicker} from 'react-dates'
import {setStartDate, setEndDate} from '../actions/filters'

export class ExpenseListFilter extends React.Component {
    state = {
        focused: null
    }
    onTextChange = (value) => {
        this.props.changeFilterText(value)
    }
    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }
    onFocusChange = (focused) => {
        this.setState(() => ({focused}))
    }
    onSortChange = (value) => {
        if (value === 'date'){
            this.props.sortbydate()
        }
        else if (value === 'amount'){
            this.props.sortbyamount()
        }
    }
    render(){
        return(
            <div>
                <input type="text" value={this.props.filters.text} onChange={(e) => this.onTextChange(e.target.value)}></input>
                <select value={this.props.filters.sortBy} onChange={(e) => this.onSortChange(e.target.value)}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                startDate={this.props.filters.startDate}
                endDate={this.props.filters.endDate}
                onDatesChange={this.onDatesChange}
                focusedInput={this.state.focused}
                onFocusChange={this.onFocusChange}
                showClearDates={true}
                numberOfMonths={1}
                isOutsideRange={() => (false)}
                 />
            </div>
        )
    }
}


const MapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

const MapDispatchToProps = (dispatch) => ({
    changeFilterText: (value) => dispatch(setFilterText(value)),
    changeStartDate: (date) => dispatch(setStartDate(date)),
    changeEndDate: (date) => dispatch(setEndDate(date)),
    sortbydate: () => dispatch(sortByDate()),
    sortbyamount: () => dispatch(sortByAmount())
    
})

const ConnectedListFilter = connect(MapStateToProps, MapDispatchToProps)(ExpenseListFilter)

export default ConnectedListFilter