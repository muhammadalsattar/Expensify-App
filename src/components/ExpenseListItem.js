import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import moment from 'moment';
import numeral from 'numeral'

export function ExpenseListItem({ id, description, amount, createdAt }) {
  return (
    <div>
      <Link to={`/edit/${id}`}>
        <h3>{description}</h3>
        <p>{numeral(amount/100).format('$0,0.00')} - {moment(createdAt).format('MMMM Do, YYYY')}</p>
      </Link>
    </div>
  );
}


const ConnectedExpenseItem = connect()(ExpenseListItem)

export default ConnectedExpenseItem;