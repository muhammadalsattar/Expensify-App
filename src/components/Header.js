import React from "react"
import { connect } from "react-redux";
import { NavLink } from "react-router-dom";
import { startLogout } from "../actions/auth";

export const Header = ({logout}) => (
    <div>
        <h1>Expensify</h1>
        <div>
            <NavLink activeClassName="is-active" exact to="/dashboard">Home</NavLink>
            <NavLink activeClassName="is-active" to="/create">Add expense</NavLink>
        </div>
        <button onClick={logout}>Logout</button>
    </div>
)

const mapDispatchToProps = (dispatch)=>{
    return {
        logout: ()=>{dispatch(startLogout)}
    }
}


export default connect(undefined, mapDispatchToProps)(Header);