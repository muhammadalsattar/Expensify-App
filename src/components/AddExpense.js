import React from "react"
import { connect } from "react-redux"
import ExpenseForm from './ExpenseForm'
import { startAddExpense } from "../actions/expenses"

export class AddExpense extends React.Component {
    onSubmit = (expense) => {
        this.props.onSubmit(expense)
        this.props.history.push('/')
    }
    render() {
        return (
            <div className="addexpense-section mx-auto">
                <h1 className="my-4">Add Expense</h1>
                <ExpenseForm onSubmit={this.onSubmit}/>
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => ({
    onSubmit: (expense) => dispatch(startAddExpense(expense))
})

const ConnectedAddExpense = connect(undefined, mapDispatchToProps)(AddExpense)
export default ConnectedAddExpense