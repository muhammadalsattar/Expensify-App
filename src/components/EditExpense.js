import React from "react"
import { connect } from "react-redux";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";
import ExpenseForm from "./ExpenseForm";


export class EditExpense extends React.Component {
    onSubmit = (expense) => {
        this.props.editexpense(this.props.expense.id, expense)
        this.props.history.push('/')
    }
    onRemove = () => {
        this.props.removeexpense(this.props.expense.id)
        this.props.history.push('/')
    }

    render() {
        return (
            <div>
                <h1>Edit Expense</h1>
                <ExpenseForm
                expense = {this.props.expense}
                onSubmit = {this.onSubmit}
                />
                <button onClick={this.onRemove}>remove</button>
            </div>
        )
    }
}


const MapStateToProps = (state, props) => ({
    expense: state.expenses.find(expense => (expense.id === props.match.params.id))
})

const MapDispatchToProps = (dispatch, props)=> ({
    editexpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    removeexpense: (data) => dispatch(startRemoveExpense(data))
})


export default connect(MapStateToProps, MapDispatchToProps)(EditExpense);