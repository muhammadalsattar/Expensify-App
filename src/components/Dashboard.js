import React from "react"
import ExpenseList from './ExpenseList'
import ConnectedListFilter from "./ExpenseListFilter";
import ConnectedExpensesTotal from "./ExpensesTotal";

export const Dashboard = () => (
    <div>
        <ConnectedExpensesTotal/>
        <ConnectedListFilter/>
        <ExpenseList />
    </div>
)

export default Dashboard;