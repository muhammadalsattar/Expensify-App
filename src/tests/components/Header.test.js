import React from "react";
import { Header } from "../../components/Header";
// import ReactShallowRenderer from 'react-test-renderer/shallow'
import {shallow} from 'enzyme'

let wrapper, startLogout;

beforeEach(()=>{
    startLogout = jest.fn()
    wrapper = shallow(<Header logout={startLogout}/>)
})

test('Should render Header component', () => {
    // const renderer = new ReactShallowRenderer();
    // renderer.render(<Header/>)
    // expect(renderer.getRenderOutput()).toMatchSnapshot()
    wrapper = shallow(<Header/>)
    expect(wrapper).toMatchSnapshot()
})

test('Should call logout function', ()=>{
    wrapper.find('button').simulate('click');
    expect(startLogout).toHaveBeenCalled()
})