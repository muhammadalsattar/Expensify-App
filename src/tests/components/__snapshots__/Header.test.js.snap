// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Header component 1`] = `
<div>
  <h1>
    Expensify
  </h1>
  <div>
    <NavLink
      activeClassName="is-active"
      exact={true}
      to="/dashboard"
    >
      Home
    </NavLink>
    <NavLink
      activeClassName="is-active"
      to="/create"
    >
      Add expense
    </NavLink>
  </div>
  <button>
    Logout
  </button>
</div>
`;
